src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketJobLinkActionFactory.java

```diff
@@ -0,0 +1,118 @@
+import com.atlassian.bitbucket.jenkins.internal.config.BitbucketServerConfiguration;
+import com.google.common.annotations.VisibleForTesting;
+import hudson.Extension;
+import hudson.model.Action;
+import hudson.model.FreeStyleProject;
+import hudson.model.ItemGroup;
+import hudson.model.Job;
+import hudson.scm.SCM;
+import hudson.util.FormValidation;
+import jenkins.model.TransientActionFactory;
+import org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;
+
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import java.util.*;
+
+@Extension
+public class BitbucketJobLinkActionFactory extends TransientActionFactory<Job> {
+
+    @Inject
+    private BitbucketPluginConfiguration bitbucketPluginConfiguration;
+    @Inject
+    private BitbucketScmFormValidationDelegate formValidation;
+
+    public BitbucketJobLinkActionFactory() { }
+
+    @Inject
+    public BitbucketJobLinkActionFactory(BitbucketPluginConfiguration bitbucketPluginConfiguration,
+                                         BitbucketScmFormValidationDelegate formValidation) {
+        this.bitbucketPluginConfiguration = bitbucketPluginConfiguration;
+        this.formValidation = formValidation;
+    }
+
+    @Nonnull
+    @Override
+    public Collection<? extends Action> createFor(@Nonnull Job target) {
+        Optional<BitbucketSCMRepository> maybeRepository = getBitbucketSCMRepository(target);
+        if (!maybeRepository.isPresent()) {
+            return Collections.emptySet();
+        }
+        BitbucketSCMRepository bitbucketRepository = maybeRepository.get();
+        String serverId = Objects.toString(bitbucketRepository.getServerId(), "");
+        String credentialsId = Objects.toString(bitbucketRepository.getCredentialsId(), "");
+
+        Optional<BitbucketServerConfiguration> maybeConfig = bitbucketPluginConfiguration.getServerById(serverId);
+        FormValidation configValid = FormValidation.aggregate(Arrays.asList(
+                maybeConfig.map(BitbucketServerConfiguration::validate).orElse(FormValidation.error("Config not present")),
+                formValidation.doCheckProjectName(target, serverId, credentialsId, bitbucketRepository.getProjectName()),
+                formValidation.doCheckRepositoryName(target, serverId, credentialsId, bitbucketRepository.getProjectName(), bitbucketRepository.getRepositoryName())
+        ));
+
+        if (configValid.kind == FormValidation.Kind.ERROR) {
+            return Collections.emptySet();
+        }
+
+        String url = maybeConfig.get().getBaseUrl() +
+                     "/projects/" +
+                     bitbucketRepository.getProjectKey() +
+                     "/repos/" +
+                     bitbucketRepository.getRepositorySlug();
+        return Collections.singleton(BitbucketExternalLink.createDashboardLink(url, target));
```

Comment:
Does this make all links point to the repository in Bitbucket Server? In [](https://issues.jenkins-ci.org/browse/JENKINS-63070), I hoped that each Jenkins job corresponding to a branch would link to that branch in Bitbucket Server, and likewise with pull requests. Is that implemented in some other method?

Reply:
Thanks for the comments Kalle, I will address these in a separate PR. I didn't include the main workflow project because it's possible to have multiple branch sources (and I didn't want to flood the project page with links)- chatting with team mates, it seems like this is possible, but not a solution we support so I'll make the changes as requested.